test/OpenTelemetry.Instrumentation.AspNetCore.Tests/BasicTests.cs

```diff
@@ -694,6 +695,136 @@ public async Task ShouldExportActivityWithOneOrMoreExceptionFilters(int mode)
             AssertException(exportedItems);
         }
 
+        [Fact(Skip = "Pending Changes https://github.com/open-telemetry/opentelemetry-dotnet/issues/3495")]
+        public async Task DiagnosticSourceCustomCallbacksAreReceivedOnlyForSubscribedEvents()
+        {
+            int numberOfCustomCallbacks = 0;
+            string expectedCustomEventName = "Microsoft.AspNetCore.Mvc.BeforeAction";
+            string actualCustomEventName = null;
+            void ConfigureTestServices(IServiceCollection services)
+            {
+                this.tracerProvider = Sdk.CreateTracerProviderBuilder()
+                    .AddAspNetCoreInstrumentation(new AspNetCoreInstrumentation(
+                        new TestHttpInListener(new AspNetCoreInstrumentationOptions())
+                        {
+                            OnCustomCallback = (name, activity, payload) =>
+                            {
+                                actualCustomEventName = name;
+                                numberOfCustomCallbacks++;
+                            },
+                        }))
+                    .Build();
+            }
+
+            // Arrange
+            using (var client = this.factory
+                .WithWebHostBuilder(builder =>
+                    builder.ConfigureTestServices(ConfigureTestServices))
+                .CreateClient())
+            {
+                using var request = new HttpRequestMessage(HttpMethod.Get, "/api/values");
+
+                // Act
+                using var response = await client.SendAsync(request);
+            }
+
+            Assert.Equal(1, numberOfCustomCallbacks);
+            Assert.Equal(expectedCustomEventName, actualCustomEventName);
+        }
+
+        [Fact]
+        public async Task DiagnosticSourceExceptionCallbackIsReceivedForUnHandledException()
+        {
+            int numberOfExceptionCallbacks = 0;
+            void ConfigureTestServices(IServiceCollection services)
+            {
+                this.tracerProvider = Sdk.CreateTracerProviderBuilder()
+                    .AddAspNetCoreInstrumentation(new AspNetCoreInstrumentation(
+                        new TestHttpInListener(new AspNetCoreInstrumentationOptions())
+                        {
+                            OnExceptionCallback = (activity, payload) =>
+                            {
+                                numberOfExceptionCallbacks++;
+                            },
```

Comment:
Not a blocker for this PR, but...

Just like OnCustom, ideally we'd capture the event name for OnException and assert we got the expected one.  But we don't have it.

Seems the AspNetCore and Http instrumentation are the only ones that use OnException. Maybe we could migrate these usages to OnCustom in another PR. Other option I guess would be to pass the event name to OnException.

Reply:
Yes - I was thinking to get rid of OnStart, OnStop and OnException as per discussion [here](https://github.com/open-telemetry/opentelemetry-dotnet/pull/3519#discussion_r936091699). Wanted to ask you and @cijothomas if you see any issues if we do that?.

Reply:
I'm in full support of ripping this bandaid off 🩹 sooner than later. This PR just made it look like you were considering doing things incrementally.

Reply:
Yes - wanted to keep that change separate as that will involve SDK changes.